src/xrpld/app/tx/detail/ConfidentialSend.cpp

```diff
@@ -0,0 +1,211 @@
+        return tecNO_PERMISSION;
+
+    // Check destination's MPToken
+    auto const sleDestinationMPToken =
+        ctx.view.read(keylet::mptoken(mptIssuanceID, destination));
+    if (!sleDestinationMPToken)
+        return tecOBJECT_NOT_FOUND;
+
+    // Check destination's MPToken has necessary fields for confidential send
+    if (!sleDestinationMPToken->isFieldPresent(sfHolderElGamalPublicKey) ||
+        !sleDestinationMPToken->isFieldPresent(sfConfidentialBalanceInbox) ||
+        !sleDestinationMPToken->isFieldPresent(sfIssuerEncryptedBalance))
+        return tecNO_PERMISSION;
+
+    // todo: check zkproof. equality proof and range proof, combined or separate
+    // TBD. TER const terProof = verifyConfidentialSendProof(
+    //     ctx.tx[sfZKProof],
+    //     (*sleSender)[sfConfidentialBalanceSpending],
+    //     ctx.tx[sfSenderEncryptedAmount],
+    //     ctx.tx[sfDestinationEncryptedAmount],
+    //     ctx.tx[sfIssuerEncryptedAmount],
+    //     (*sleSender)[sfHolderElGamalPublicKey],
+    //     (*sleDestination)[sfHolderElGamalPublicKey],
+    //     (*sleIssuance)[sfIssuerElGamalPublicKey],
+    //     (*sleSender)[~sfConfidentialBalanceVersion].value_or(0),
+    //     ctx.tx.getTransactionID()
+    // );
+
+    // if (!isTesSuccess(terProof))
+    //     return tecBAD_PROOF;
+
+    return tesSUCCESS;
+}
+
+TER
+ConfidentialSend::doApply()
+{
+    auto const mptIssuanceID = ctx_.tx[sfMPTokenIssuanceID];
+    auto const account = ctx_.tx[sfAccount];
+    auto const destination = ctx_.tx[sfDestination];
+
+    auto sleSender = view().peek(keylet::mptoken(mptIssuanceID, account));
+    auto sleDestination =
+        view().peek(keylet::mptoken(mptIssuanceID, destination));
+
+    if (!sleSender || !sleDestination)
+        return tecINTERNAL;
+
+    Slice const senderEc = ctx_.tx[sfSenderEncryptedAmount];
+    Slice const destEc = ctx_.tx[sfDestinationEncryptedAmount];
+    Slice const issuerEc = ctx_.tx[sfIssuerEncryptedAmount];
+
+    // Substract from sender's spending balance
+    {
+        Slice const curSpending = (*sleSender)[sfConfidentialBalanceSpending];
+        Buffer newSpending(ecGamalEncryptedTotalLength);
+
+        if (TER const ter =
+                homomorphicSubtract(curSpending, senderEc, newSpending);
+            !isTesSuccess(ter))
+            return tecINTERNAL;
+
+        (*sleSender)[sfConfidentialBalanceSpending] = newSpending;
+    }
+
+    // Substract from issuer's balance
+    {
+        Slice const curIssuerEnc = (*sleSender)[sfIssuerEncryptedBalance];
+        Buffer newIssuerEnc(ecGamalEncryptedTotalLength);
+
+        if (TER const ter =
+                homomorphicSubtract(curIssuerEnc, issuerEc, newIssuerEnc);
+            !isTesSuccess(ter))
+            return tecINTERNAL;
+
+        (*sleSender)[sfIssuerEncryptedBalance] = newIssuerEnc;
+    }
+
+    // Increment version
+    (*sleSender)[sfConfidentialBalanceVersion] =
+        (*sleSender)[sfConfidentialBalanceVersion] + 1;
```

Comment:
```suggestion
    (*sleSender)[sfConfidentialBalanceVersion] =
        (*sleSender)[~sfConfidentialBalanceVersion].value_or(0) + 1;
```
let's do this for safety